File: src/cli/query/get.ts
import { Command, ExactArgs, ref } from "vipvot";
import { printJson, printError } from "../../lib/output.ts";
import { getMongoClient, closeAllClients } from "../../mongo/client.ts";
import { findById } from "../../mongo/query.ts";
import { enhanceErrorMessage } from "../../lib/errors.ts";
import { parseJson } from "../../lib/parse-json.ts";
import { resolveConnectionAlias } from "../_globals.ts";

export function buildGetCommand(): Command {
  const idType = ref<string>("");
  const projection = ref<string>("");

  const cmd = Command({
    use: "get <database> <collection> <id>",
    short: "Get a single document by _id",
    args: ExactArgs(3),
    run: async (_cmd, args) => {
      const [database, collection, id] = args as [string, string, string];
      try {
        if (idType.value && !["objectid", "string", "number"].includes(idType.value)) {
          throw new Error(`Invalid --type: "${idType.value}". Valid: objectid, string, number`);
        }

        const { client } = await getMongoClient(resolveConnectionAlias());
        const projectionDoc = projection.value
          ? parseJson(projection.value, "projection")
          : undefined;
        const doc = await findById(client, {
          dbName: database,
          collName: collection,
          rawId: id,
          idType: idType.value || undefined,
          projection: projectionDoc,
        });

        if (!doc) {
          throw new Error(`Document not found: _id=${id} in ${database}.${collection}`);
        }

        printJson({
          database,
          collection,
          fieldCount: Object.keys(doc).length,
          document: doc,
        });
      } catch (err) {
        printError(
          err instanceof Error
            ? enhanceErrorMessage(err, { database, collection })
            : "Failed to get document",
        );
      } finally {
        await closeAllClients();
      }
    },
  });

  cmd
    .flags()
    .stringVarP(
      idType,
      "type",
      "",
      "",
      "Force ID type: objectid, string, number (auto-detected by default)",
    );
  cmd
    .flags()
    .stringVarP(
      projection,
      "projection",
      "",
      "",
      'Field projection (e.g. {"name": 1, "email": 1})',
    );

  return cmd;
}

File: internal/cli/query/get.go
package query

import (
	"fmt"

	"github.com/spf13/cobra"

	"github.com/shhac/agent-mongo/internal/cli/shared"
	"github.com/shhac/agent-mongo/internal/mongo"
	"github.com/shhac/agent-mongo/internal/output"
)

func registerGet(parent *cobra.Command, globals func() *shared.GlobalFlags) {
	var idType, projection string

	cmd := &cobra.Command{
		Use:   "get <database> <collection> <id>",
		Short: "Get a single document by _id",
		Args:  cobra.ExactArgs(3),
		RunE: func(_ *cobra.Command, args []string) error {
			g := globals()
			ref := mongo.Ref{DB: args[0], Collection: args[1]}
			id := args[2]

			if idType != "" && idType != "objectid" && idType != "string" && idType != "number" {
				return fmt.Errorf("Invalid --type: %q. Valid: objectid, string, number", idType)
			}
			projectionDoc, err := parseOptionalDoc(projection, "projection")
			if err != nil {
				return err
			}

			return shared.WithSessionRef(g, ref, func(ctx shared.SessionCtx) error {
				doc, err := ctx.Session.FindByID(ctx.Ctx, mongo.FindByIDOpts{
					Ref:        ref,
					RawID:      id,
					IDType:     idType,
					Projection: projectionDoc,
				})
				if err != nil {
					return err
				}
				if doc == nil {
					return fmt.Errorf("Document not found: _id=%s in %s.%s", id, ref.DB, ref.Collection)
				}
				return output.PrintResult(map[string]any{
					"database":   ref.DB,
					"collection": ref.Collection,
					"fieldCount": len(doc),
					"document":   doc,
				})
			})
		},
	}

	cmd.Flags().StringVar(&idType, "type", "",
		"Force ID type: objectid, string, number (auto-detected by default)")
	cmd.Flags().StringVar(&projection, "projection", "", `Field projection (e.g. {"name": 1, "email": 1})`)
	parent.AddCommand(cmd)
}
